src/main/java/fi/aalto/cs/apluscourses/intellij/actions/PocAction.java

```diff
@@ -0,0 +1,173 @@
+import com.intellij.openapi.actionSystem.DataContext;
+import com.intellij.openapi.actionSystem.ex.AnActionListener;
+import com.intellij.openapi.fileEditor.FileEditor;
+import com.intellij.openapi.fileEditor.FileEditorManager;
+import com.intellij.openapi.fileEditor.FileEditorManagerListener;
+import com.intellij.openapi.project.DumbAwareAction;
+import com.intellij.openapi.roots.ProjectRootManager;
+import com.intellij.openapi.vfs.VirtualFile;
+import com.intellij.openapi.vfs.VirtualFileManager;
+import com.intellij.openapi.vfs.newvfs.BulkFileListener;
+import com.intellij.openapi.vfs.newvfs.events.VFileContentChangeEvent;
+import com.intellij.openapi.vfs.newvfs.events.VFileEvent;
+import com.intellij.openapi.vfs.newvfs.events.VFilePropertyChangeEvent;
+import com.intellij.refactoring.listeners.RefactoringEventData;
+import com.intellij.refactoring.listeners.RefactoringEventListener;
+import com.intellij.util.concurrency.annotations.RequiresReadLock;
+import com.intellij.util.messages.MessageBusConnection;
+import fi.aalto.cs.apluscourses.intellij.model.APlusProject;
+
+import java.util.List;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public class PocAction extends DumbAwareAction {
+
+  @NotNull
+  private APlusProject project;
+
+  @Override
+  public void actionPerformed(@NotNull AnActionEvent e) {
+    project = new APlusProject(e.getProject());
+    //explore what events can be listened to and register related listeners
+
+    //1 - Listen for file save event (and 2 - access contents)
+    //3 - Listen for file rename event
+    FileListener fileListener = new FileListener();
+    fileListener.registerListeners();
+
+    // 4 - Existing IntelliJ's buttons clicked e.g. debug, run
+    APlusActionListener myAnActionListener = new APlusActionListener();
+    myAnActionListener.registerListener();
+
+    // 5 - File opened! Also, get the file's path.
+    OpenListener openListener = new OpenListener("/GoodStuff/o1/goodstuff/Category.scala");
+    openListener.registerListeners();
+
+    //Refactoring element listener
+    RefactoringListener refactoringListener = new RefactoringListener();
+    refactoringListener.registerListeners();
+  }
+
+  private class APlusActionListener implements AnActionListener {
+
+    public void registerListener() {
+      MessageBusConnection messageBusConnection = project.getMessageBus().connect();
+      messageBusConnection.subscribe(AnActionListener.TOPIC, this);
+    }
+
+    @Override
+    public void beforeActionPerformed(@NotNull AnAction action, @NotNull DataContext dataContext,
+                                      @NotNull AnActionEvent event) {
+      System.out.println("Action: " + action.getTemplateText()
+              + " ActionEvent: " + event.getPlace());
+    }
+  }
+
+  private class RefactoringListener implements RefactoringEventListener {
+    @RequiresReadLock
+    public synchronized void registerListeners() {
```

Comment:
`APlusActionListener;;registerListener` doesn't have `@RequiresReadLock`? Is that intentional that there's a difference (they're, after all, identical except for topic). If not, do you know which one is the correct one?

Reply:
Not intentional, I missed adding the `@RequiresReadLock` to `APlusActionListener`.